src/python/WMCore/Services/DDM/DDM.py

```diff
@@ -0,0 +1,286 @@
+        """
+
+        for key in list(self):
+            if self[key] is None:
+                self.pop(key)
+
+    def isEqual(self, ddmReq, exclude=None):
+        """
+        _isEqual_
+
+        Equivalence method.
+        :exclude:
+            a key to be excluded from the comparison.
+        :return:
+            Bool
+        """
+        # first we check the apis - we cannot compare apples to oranges:
+        if not self.api == ddmReq.api:
+            return False
+
+        # in case no key have been excluded it is a normal dict comparison
+        if not exclude:
+            return self == ddmReq
+
+        # The final check: we iterate over the two objects separately in order
+        # to be sure we do not leave any extra key which is present in one of
+        # the objects but not in the other
+        for key in self:
+            if key in exclude:
+                continue
+            try:
+                if ddmReq[key] != self[key]:
+                    return False
+            except KeyError:
+                return False
+
+        for key in ddmReq:
+            if key in exclude:
+                continue
+            try:
+                if ddmReq[key] != self[key]:
+                    return False
+            except KeyError:
+                return False
+        return True
+
+
+class DDM(Service):
+    """
+    Class which provides client APIs to the DDM service.
+
+    Usage example:
+        api = 'copy'
+        ddmReq = DDMReqTemplate(api)
+        ddm = DDM()
+        ddm.makeRequest(ddmReq, api)
+    """
+
+    def __init__(self, url=None, logger=None, configDict=None):
+        """
+        configDict:
+            Dictionary with parameters that are passed to the super class.
+        """
+        url = url or "https://dynamo.mit.edu"
+        configDict = configDict or {}
+        configDict.setdefault('endpoint', url)
+        configDict.setdefault('cacheduration', 1)  # in hours
+        configDict.setdefault('accept_type', 'application/json')
+        configDict.setdefault('content_type', 'application/json')
+        configDict['logger'] = logger if logger else logging.getLogger()
+        super(DDM, self).__init__(configDict)
+        self['logger'].debug("Initializing DDM with url: %s", self['endpoint'])
+
+    def _getResult(self, ddmReq, apiUrl, callname=""):
+        """
+        Either fetch data from the cache file or query the data-service
+        :param metricNumber: a number corresponding to the SSB metric
+        :return: a dictionary
+
+        """
+
+        # TODO: to make this cache not per call but per DDM object instance
+        cachedApi = "%s.json" % callname
+
+        # striping the None values from the request before sending it to dynamo
+        ddmReq.strip()
+        msg = "INFO: sending data to DDM: "
+        msg += "API: {}: ddmReq: {}".format(ddmReq.api, pformat(ddmReq))
+        self['logger'].info(msg)
```

Comment:
Maybe this logging level has to be debug before we merge it?

Reply:
agreed

Reply:
Done